src/plugins/_manager.ts

```diff
@@ -0,0 +1,171 @@
+import * as child from 'child_process';
+import * as os from 'os';
+import * as path from 'path';
+import * as url from 'url';
+import * as fs from 'fs-extra';
+
+const MODULE_NOT_FOUND_ERROR_CODE = 'MODULE_NOT_FOUND'; // TODO is there a known constant we can use here?
+const MODULE_VERSION_MISMATCH_ERROR_CODE = 'MODULE_VERSION_MISMATCH';
+
+/**
+ * Error indicating a module was found but with an unexpected version.
+ */
+export class ModuleVersionMismatchError extends Error {
+
+  public readonly code: string = MODULE_VERSION_MISMATCH_ERROR_CODE;
+
+  constructor(pkg: string, expected: string, actual: string) {
+    super(`Version mismatch for package ${pkg}. Found ${actual} but requested ${expected}`);
+  }
+}
+
+export interface Plugin {
+
+  readonly instance: any;
+  readonly class: string;
+  readonly module: any;
+  readonly pkg: string;
+  readonly version: string;
+  readonly path: string;
+
+}
+
+export interface Package {
+
+  readonly module: any;
+  readonly pkg: string;
+  readonly version: string;
+  readonly path: string;
+
+}
+
+export interface PluginManagerLoadOptions {
+  readonly pkg: string;
+  readonly version: string;
+  readonly class: string;
+  readonly installEnv?: { [key: string]: any };
+  readonly properties?: { [key: string]: any };
+}
+
+export class PluginManager {
+
+  private readonly dir: string;
+
+  constructor(dir: string) {
+    this.dir = dir;
+  }
```

Comment:
Why not just:

```suggestion
  constructor(private readonly dir: string) {}
```

IMHO the only reasons to do the assignment explicitly is:
1. You're doing some validations pre-assignment
2. You're using a `#privateField`, and these can't be declared in the constructor signature

Reply:
I still have some java mannerisms :)